newsService/src/main/java/com/kovan/app/service/NewsService.java

```diff
@@ -34,50 +39,89 @@ public class NewsService {
+                        throw new NewsRetrievalException("Failed to parse news data from API response.", e);
+                    }
+                })
+                .takeWhile(newsDto -> PAGE_SIZE.addAndGet(DEFAULT_PAGE_SIZE) < newsDto.getTotalResults())
+                .toList();
+
+        boolean isYesterdayInDb = isNewsAlreadyInDb(yesterday.toString());
+        boolean isTodayInDb = isNewsAlreadyInDb(today.toString());
+
+        Map<Boolean, List<NewsDto.Article>> partitionedArticles = articles.get().stream()
+                .takeWhile(article -> !parseDate(article.getPublishedAt()).isBefore(yesterday))
+                .collect(partitioningBy(article -> parseDate(article.getPublishedAt()).equals(yesterday)));
+
+        newsArticles = partitionedArticles.get(true);
+        futureArticles = partitionedArticles.get(false);
+
+        if (newsArticles.isEmpty() && futureArticles.isEmpty() && !isYesterdayInDb) {
+            NewsDto emptyNews = NewsDto.builder().totalResults(0)
+                    .publishedAt(yesterday.toString()).status("fail").build();
+           return singletonList(service.saveNewsInDb(emptyNews));
+        } else {
+            savedYesterdayNewsDto = handleNewsSaving(isYesterdayInDb,newsArticles,yesterday);
+            savedTodayNewsDto = handleNewsSaving(isTodayInDb,futureArticles,today);
+        }
 
-    public NewsDto getTopHeadlines() {
-
-        List<NewsDto.Article> articles = new ArrayList<>();
-        int pageSize = 0;
-        int page = 1;
-        NewsDto newsDto;
-
-        do {
-            pageSize += 20;
-            String apiUrl = buildUrl(page++);
-            String response;
-            try {
-                response = restTemplate.getForObject(apiUrl, String.class);
-                newsDto = objectMapper.readValue(response, NewsDto.class);
-                articles.addAll(newsDto.getArticles());
-            } catch (JsonProcessingException e) {
-                throw new NewsRetrievalException("Failed to parse news data from API response.", e);
-            }
-
-        } while (pageSize <= newsDto.getTotalResults());
-
-        LocalDate dateOnly = LocalDate.parse(
-                articles.getFirst().getPublishedAt(),
-                DateTimeFormatter.ISO_DATE_TIME
-        );
-
-        if (!isNewsAlreadyInDb(dateOnly.toString())) {
-            NewsDto finalDto = NewsDto.builder()
-                    .articles(articles)
-                    .totalResults(articles.size())
-                    .publishedAt(dateOnly.toString())
-                    .build();
-
-            newsDto = service.saveNewsInDb(finalDto);
+       return of(savedTodayNewsDto, savedYesterdayNewsDto)
+               .flatMap(Optional::stream)
+               .toList();
+    }
+
+    private Optional<NewsDto> handleNewsSaving(boolean isInDb, List<NewsDto.Article> articles,LocalDate date) {
+        if(!isInDb && !articles.isEmpty()){
+           return Optional.of(saveNews(articles,date));
```

Comment:
do static import

Reply:
Raises ambiguity between Optional and Stream.